rasgotransforms/rasgotransforms/jinja.py

```diff
@@ -0,0 +1,60 @@
+from jinja2 import Environment
+from typing import Callable, Dict, Optional, Tuple, Union
+from itertools import combinations, permutations, product
+import re
+
+JINJA_ENV = Environment(extensions=['jinja2.ext.do', 'jinja2.ext.loopcontrols'])
+
+
+def cleanse_template_symbol(symbol: str) -> str:
+    symbol = str(symbol)
+    symbol = symbol.strip()
+    symbol = symbol.replace(' ', '_').replace('-', '_')
+    symbol = symbol.upper()
+    symbol = re.sub('[^A-Z0-9_]+', '', symbol)
+    symbol = '_' + symbol if symbol[0].isdecimal() or not symbol else symbol
+
+    return symbol
+
+
+def raise_exception(message: str) -> None:
+    raise Exception(message)
+
+
+def render(
+    source_code: str,
+    source_table: str,
+    arguments: dict,
+    source_columns: Optional[Union[Dict[str, Dict[str, str]], Callable]] = None,
+    run_query: Optional[Callable] = None,
+) -> str:
+    """
+    Given the source code of a jinja template, its arguments, and metadata about the source table,
+    this renders the transform sql.
+
+    Source columns can be a mapping of table names to columns (e.g. {table_name: {column_name: column_type}})
+    or a function which returns a mapping of columns names to column types given a table fqtn
+    (e.g. get_columns(fqtn): return {column_name: column_type}
+
+    Run Query is a function which takes arbitrary sql and executes it agains the data warehouse. The data must be
+    returned as a Pandas dataframe
+    """
+    arguments['source_table'] = source_table
+    if source_columns and type(source_columns) is dict:
```

Comment:
Why allow dict _or_ a function? Why not just keep it simple and make it always a function (that just returns the correct dict)

For transparency, this is what I was thinking: 
https://github.com/rasgointelligence/render-transforms/blob/master/lambda_function.py#L46

Reply:
+1, but also why accept source_columns at all? Just let them pass in get_columns as one of the functions in the override dict as suggested above.
We're going to find 10 more macros we think every transform needs if this package gets wider adoption. We should draw a line early and say all macros (functions) get passed in a single arg to keep the `render` function signature clean

Reply:
I will add the override dict for other macros. My thoughts for adding the source_columns arg and accepting a dict or a function were that if we had a function which returns the rendered sql and the inferred output columns then we could just pass those output columns into the function again to render another transform, so chaining lots of transforms is really easy like this:

```
transform1_sql, output_columns = render(..., infer_columns=True)
transform2_sql, output_columns = render(..., source_columns=output_columns, infer_columns=True)
```

The reason I take a function and not just a dict is so that our api or other clients that work similarily can pass in a 'get_columns' function.

If accepting the source_columns like this doesn't seem useful, I can take it out.